metadata-ingestion/src/datahub/ingestion/source/dbt/dbt_common.py

```diff
@@ -707,6 +850,182 @@ def filter_nodes(self, all_nodes: List[DBTNode]) -> List[DBTNode]:
+
+            # Otherwise, load the schema from the dbt catalog.
+            # Note that this might get the casing wrong relative to DataHub, but
+            # has a more up-to-date column list.
+            if node.columns and (
+                not schema_fields or self.config.convert_column_urns_to_lowercase
+            ):
+                schema_fields = [
+                    SchemaField(
+                        fieldPath=column.name.lower()
+                        if self.config.convert_column_urns_to_lowercase
+                        else column.name,
+                        type=column.datahub_data_type
+                        or SchemaFieldDataType(type=NullTypeClass()),
+                        nativeDataType=column.data_type,
+                    )
+                    for column in node.columns
+                ]
+
+            # Add the node to the schema resolver, so that we can get column
+            # casing to match the upstream platform.
+            added_to_schema_resolver = False
+            if target_node_urn and schema_fields:
+                schema_resolver.add_raw_schema_info(
+                    target_node_urn, self._to_schema_info(schema_fields)
+                )
+                added_to_schema_resolver = True
+
+            # Run sql parser to infer the schema + generate column lineage.
+            sql_result = None
+            if node.compiled_code:
+                try:
+                    # Add CTE stops based on the upstreams list.
+                    preprocessed_sql = detach_ctes(
+                        node.compiled_code,
+                        platform=schema_resolver.platform,
+                        cte_mapping={
+                            cte_name: upstream_node.get_fake_ephemeral_table_name()
+                            for upstream_node in [
+                                all_nodes_map[upstream_node_name]
+                                for upstream_node_name in node.upstream_nodes
+                                if upstream_node_name in all_nodes_map
+                            ]
+                            if upstream_node.is_ephemeral_model()
+                            for cte_name in _get_dbt_cte_names(
+                                upstream_node.name, schema_resolver.platform
+                            )
+                        },
+                    )
+                except Exception as e:
+                    sql_result = SqlParsingResult.make_from_error(e)
+                else:
+                    sql_result = sqlglot_lineage(
+                        preprocessed_sql, schema_resolver=schema_resolver
+                    )
+
+            # Save the column lineage.
+            if self.config.include_column_lineage and sql_result:
+                # We only save the debug info here. We're report errors based on it later, after
+                # applying the configured node filters.
+                node.cll_debug_info = sql_result.debug_info
```

Comment:
Attaching lineage info to nodes is different than we do most other sources. How come you're doing it this way here?

Reply:
lineage is determined by the view definition and the schema of the upstreams. in most other sources, we have the schemas available, but in the case of dbt ephemeral models, we have to infer the schemas. that means we need to do it topographical order, so it's easier to do it all in one go